src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs

```diff
@@ -452,22 +452,25 @@ public StringSegment Subsegment(int offset, int length)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public int IndexOf(char c, int start, int count)
         {
-            int offset = Offset + start;
+            int index = -1;
 
-            if (!HasValue || start < 0 || (uint)offset > (uint)Buffer.Length)
+            if (HasValue)
             {
-                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.start);
-            }
+                if (start < 0 || start > Length)
+                {
+                    ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.start);
+                }
 
-            if (count < 0)
-            {
-                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.count);
-            }
+                if (count < 0 || start + count > Length)
```

Comment:
`start + count` could overflow, so become a negative sum. To be on the safe side:
```suggestion
                if (count < 0 || count > Length - start)
```
and to save one comparison:
```suggestion
                if ((uint)count > (uint)(Length - start))
```

Similar below.